OrbitGl/LiveFunctionsController.cpp

```diff
@@ -0,0 +1,150 @@
+      id_with_min_timestamp = it.first;
+    }
+    next_boxes.insert(std::make_pair(it.first, box));
+  }
+
+  // We only want to commit to the new boxes when all boxes can be moved.
+  current_textboxes_ = next_boxes;
+  id_to_select_ = id_with_min_timestamp;
+  Move();
+  return true;
+}
+
+bool LiveFunctionsController::OnAllPreviousButton() {
+  absl::flat_hash_map<uint64_t, const TextBox*> next_boxes;
+  uint64_t id_with_min_timestamp = 0;
+  uint64_t min_timestamp = std::numeric_limits<uint64_t>::max();
+  for (auto it : function_iterators_) {
+    Function* function = it.second;
+    auto function_address =
+        FunctionUtils::GetAbsoluteAddress(*function);
+    const TextBox* current_box = current_textboxes_.find(it.first)->second;
+    const TextBox* box = GCurrentTimeGraph->FindPreviousFunctionCall(
+        function_address, current_box->GetTimer().m_End);
+    if (box == nullptr) {
+      return false;
+    }
+    if (box->GetTimer().m_Start < min_timestamp) {
+      min_timestamp = box->GetTimer().m_Start;
+      id_with_min_timestamp = it.first;
+    }
+    next_boxes.insert(std::make_pair(it.first, box));
+  }
+
+  // We only want to commit to the new boxes when all boxes can be moved.
+  current_textboxes_ = next_boxes;
+  id_to_select_ = id_with_min_timestamp;
+  Move();
+  return true;
+}
+
+void LiveFunctionsController::OnNextButton(uint64_t id) {
+  auto function_address =
+        FunctionUtils::GetAbsoluteAddress(*(function_iterators_[id]));
+  const TextBox* text_box = GCurrentTimeGraph->FindNextFunctionCall(
+      function_address, current_textboxes_[id]->GetTimer().m_End);
+  // If text_box is nullptr, then we have reached the right end of the timeline.
+  if (text_box != nullptr) {
+    current_textboxes_[id] = text_box;
+  }
+  id_to_select_ = id;
+  Move();
+}
+void LiveFunctionsController::OnPreviousButton(uint64_t id) {
+  auto function_address =
+        FunctionUtils::GetAbsoluteAddress(*(function_iterators_[id]));
+  const TextBox* text_box = GCurrentTimeGraph->FindPreviousFunctionCall(
+      function_address, current_textboxes_[id]->GetTimer().m_End);
+  // If text_box is nullptr, then we have reached the left end of the timeline.
+  if (text_box != nullptr) {
+    current_textboxes_[id] = text_box;
+  }
+  id_to_select_ = id;
+  Move();
+}
+
+void LiveFunctionsController::OnDeleteButton(uint64_t id) {
+  current_textboxes_.erase(id);
+  function_iterators_.erase(id);
+  // If we erase the iterator that was last used by the user, then
+  // we need to switch last_id_pressed_ to an existing id.
+  if (id == id_to_select_ && !current_textboxes_.empty()) {
+    id_to_select_ = current_textboxes_.begin()->first;
+  } else if (current_textboxes_.empty()) {
+    // TODO: Not sure this is a good idea...
+    id_to_select_ = 0;
+  }
+  Move();
+}
+
+void LiveFunctionsController::AddIterator(Function* function) {
+  uint64_t id = next_iterator_id_;
+  ++next_iterator_id_;
+
+  auto function_address =
```

Comment:
As suggested in the comments, an easier modification to support adding an iterator in the selected text box could be something like this:

const TextBox* box = Capture::GSelectedTextBox;
if (!box){
  auto function_address = ...
  box = ....
}

Reply:
I'll try to add this, will do in a later PR, though. The main issue with just going to a random (selected) text box is that I don't know the index of that textbox (I display that in the UI, like "5 / 21"). I want to change that to just indicate roughly where in the timeline the iterator is (using the timestamps), but want to experiment with that a bit. Does that sound ok for you?

Reply:
Perfect